Fix decimal-to-integer cast behavior to match TSQL truncation

bilalkah opened the pull request:
### Description

## What is the change?

Previously, Babelfish incorrectly handled explicit conversions with ```TRY_CAST``` and ```TRY_CONVERT``` from ```sys.decimal``` to integer types (```SMALLINT```, ```INT```, ```BIGINT```). While TSQL truncates the fractional part during such conversions, Babelfish was rounding the value instead. This behavior mismatch could lead to inconsistencies in application logic when migrating from TSQL to Babelfish.

With this change, ```sys.decimal``` is explicitly handled in the ```sys.babelfish_cast_floor_*``` functions for ```SMALLINT```, ```INT```, and ```BIGINT```, ensuring the use of ```TRUNC()``` to match TSQL's truncation behavior during decimal-to-integer conversions.

## Why was the change made?

This change was made to improve Babelfish's compatibility with TSQL semantics. Accurate emulation of TSQL behavior is critical for seamless migration. Without this fix, users could encounter subtle bugs due to unexpected rounding behavior, especially in financial or index-sensitive computations.

## How was the code changed?

The following functions were updated to include ```sys.decimal``` in their ```CASE pg_typeof(arg)``` clauses:

- ```sys.babelfish_cast_floor_smallint```

- ```sys.babelfish_cast_floor_int```

- ```sys.babelfish_cast_floor_bigint```

Each of these functions now applies ```TRUNC(arg)``` for ```sys.decimal```, consistent with how TSQL behaves.

## Issues Resolved
- Corrects behavior mismatch in casting with ```TRY_CAST``` and ```TRY_CONVERT``` from sys.decimal to integer types.
- Aligns Babelfish with TSQL semantics for decimal conversion.

## Test Scenarios Covered
- Use case based
    - Verified that both ```TRY_CAST``` and ```TRY_CONVERT``` from ```DECIMAL(10,2)``` to ```SMALLINT```, ```INT```, and ```BIGINT``` correctly truncate the fractional part instead of rounding.
    - Ensured consistent behavior using both direct numeric literals and strings cast to DECIMAL.

- Boundary conditions
    - Used a consistent test input (```789.89```) to confirm that truncation results in ```789```, matching TSQL behavior.
    - Included tests with both string and numeric inputs (e.g., ```'789.89'```, ```CAST(789.89 AS DECIMAL(10,2))```).

- Arbitrary inputs
    - Covered conversions across all three target types (```SMALLINT```, ```INT```, ```BIGINT```) using decimal values.
    - Included conversion from string literals and already-cast ```DECIMAL``` values.

- Negative test cases
    - Verified ```TRY_CAST``` and ```TRY_CONVERT``` behavior with invalid or borderline inputs (e.g., ```'789.89'``` to ```SMALLINT```), ensuring they don’t throw errors and return ```NULL``` where appropriate.
    - Confirmed conversions that do not involve ```DECIMAL``` (e.g., ```'789'``` to ```BIGINT```) still work as expected.


### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

bilalkah committed:
Include sys.decimal in truncate block

Signed-off-by: Bilal Kahraman <kahramannbilal@gmail.com>

bilalkah committed:
Add test cases for decimal to integer types

Signed-off-by: Bilal Kahraman <kahramannbilal@gmail.com>

sumitj824 reviewed (commented):
Thank you @bilalkah for the findings and raising PR to fix it. The changes looks good to me, one additional consideration is handling cast from user-defined data types created using Numeric/Real/Decimal datatypes. Could you also look into extending the casting logic to properly handle those cases? You can refer the definition of `sys.bbf_xmlexist` functions to find the base data type from user-defined data type. Let us know if you need any help from us.

sumitj824 reviewed (commented):
The upgrade tests in your PR is failing because the new function definitions need to be included in the SQL upgrade script `contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--5.2.0--5.3.0.sql`.

bilalkah commented:
> The upgrade tests in your PR is failing because the new function definitions need to be included in the SQL upgrade script `contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--5.2.0--5.3.0.sql`.

Like it is done in contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.0.0--3.1.0.sql and contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.3.0--2.4.0.sql? Or should I need to insert whole functions?

sumitj824 commented:
> Like it is done in contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--3.0.0--3.1.0.sql and contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.3.0--2.4.0.sql? Or should I need to insert whole functions?

You'll need to re-create the whole functions. While those files you referenced do show partial function updates, that approach only works if you're not updating the body of the functions.

bilalkah committed:
Review fixes
- babelfish_cast functions check for base datatype
- Include changed to babelfishpg_tsql--5.2.0--5.3.0

Signed-off-by: Bilal Kahraman <kahramannbilal@gmail.com>

sumitj824 commented on contrib/babelfishpg_tsql/sql/sys_cast.sql:
```diff
@@ -13,11 +13,24 @@ STABLE;
+    arg_datatype text;
+    arg_datatype_oid oid;
+    basetype oid;
+    pltsql_quoted_identifier text;
```
Since we aren't using `pltsql_quoted_identifier` variable in the function, please remove it (applies to other similar instances).

bilalkah replied:
Done

sumitj824 commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--5.2.0--5.3.0.sql:
```diff
@@ -922,6 +922,121 @@ CREATE OR REPLACE FUNCTION sys.len(expr sys.BBF_BINARY) RETURNS INTEGER AS
+    RETURN CAST(arg AS SMALLINT);
+END; $BODY$
+LANGUAGE plpgsql
+STABLE;
```
Please remove this change, as we should only re-create those functions in upgrade script whose definitions have been modified (applies to other similar instances).

bilalkah replied:
Done

sumitj824 commented on test/JDBC/input/babel_functions_cast.sql:
```diff
@@ -306,3 +306,43 @@ select TRY_CAST(TRY_CAST(-4936.56 AS MONEY) AS varchar(10));
 select TRY_CAST(4936.56 AS smallmoney);
 GO
+
+-- Try Cast from decimal to smallint/int/bigint
```
Please add few testcases for User created data types and float data type.

bilalkah replied:
Done

sumitj824 commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--5.2.0--5.3.0.sql:
```diff
@@ -922,6 +922,121 @@ CREATE OR REPLACE FUNCTION sys.len(expr sys.BBF_BINARY) RETURNS INTEGER AS
+    END IF;
+
+    CASE arg_datatype
+        WHEN 'numeric', 'double precision', 'real', 'decimal' THEN
```
Seems like there is some inconsistency in function definition between the changes made in sys_cast.sql and upgrade script. Please make sure the function definition is identical in both files.

bilalkah replied:
Done

sumitj824 commented on contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--5.2.0--5.3.0.sql:
```diff
@@ -922,6 +922,121 @@ CREATE OR REPLACE FUNCTION sys.len(expr sys.BBF_BINARY) RETURNS INTEGER AS
+END; $BODY$
+LANGUAGE plpgsql
+STABLE;
+)
```
NIT: Seems like you added `)` by mistake, please remove it

bilalkah replied:
Done

bilalkah committed:
Review and CI fix

Signed-off-by: Bilal Kahraman <kahramannbilal@gmail.com>

bilalkah commented:
@sumitj824 The CI is completed and it is waiting for your review and approval :)

sumitj824 commented:
Task: BABEL-5930

Deepesh125 reviewed (approved):
Thank you for working on this. Please use TSQL in place of "SQL server" in the description.

sumitj824 reviewed (commented):
.
